zetaclient/zeta_supply_checker.go

```diff
@@ -0,0 +1,230 @@
+	zetaSupplyChecker := ZetaSupplyChecker{
+		stop:      make(chan struct{}),
+		ticker:    NewDynamicTicker(fmt.Sprintf("ZETASupplyTicker"), 15),
+		evmClient: make(map[int64]*ethclient.Client),
+		logger: logger.With().
+			Str("module", "ZetaSupplyChecker").
+			Logger(),
+		cfg:        cfg,
+		zetaClient: zetaClient,
+	}
+	for _, evmConfig := range cfg.GetAllEVMConfigs() {
+		if evmConfig.Chain.IsZetaChain() {
+			continue
+		}
+		client, err := ethclient.Dial(evmConfig.Endpoint)
+		if err != nil {
+			return zetaSupplyChecker, err
+		}
+		zetaSupplyChecker.evmClient[evmConfig.Chain.ChainId] = client
+	}
+
+	for chainID, _ := range zetaSupplyChecker.evmClient {
+		chain := common.GetChainFromChainID(chainID)
+		if chain.IsExternalChain() && common.IsEVMChain(chain.ChainId) && !common.IsEthereumChain(chain.ChainId) {
+			zetaSupplyChecker.externalEvmChain = append(zetaSupplyChecker.externalEvmChain, *chain)
+		}
+		if common.IsEthereumChain(chain.ChainId) {
+			zetaSupplyChecker.ethereumChain = *chain
+		}
+	}
+	balances, err := zetaSupplyChecker.zetaClient.GetGenesisSupply()
+	if err != nil {
+		return zetaSupplyChecker, err
+	}
+	tokensMintedAtBeginBlock, ok := sdkmath.NewIntFromString("200000000000000000")
+	if !ok {
+		return zetaSupplyChecker, fmt.Errorf("error parsing tokens minted at begin block")
+	}
+	zetaSupplyChecker.genesisSupply = balances.Add(tokensMintedAtBeginBlock)
+
+	logger.Info().Msgf("zeta supply checker initialized , external chains : %v ,ethereum chain :%v", zetaSupplyChecker.externalEvmChain, zetaSupplyChecker.ethereumChain)
+
+	return zetaSupplyChecker, nil
+}
+func (zs *ZetaSupplyChecker) Start() {
+	defer zs.ticker.Stop()
+	for {
+		select {
+		case <-zs.ticker.C():
+			err := zs.CheckZetaTokenSupply()
+			if err != nil {
+				zs.logger.Error().Err(err).Msgf("ZetaSupplyChecker error")
+			}
+		case <-zs.stop:
+			return
+		}
+	}
+}
+
+func (b *ZetaSupplyChecker) Stop() {
+	b.logger.Info().Msgf("ZetaSupplyChecker is stopping")
+	close(b.stop)
+}
+
+func (zs *ZetaSupplyChecker) CheckZetaTokenSupply() error {
+
+	externalChainTotalSupply := sdkmath.ZeroInt()
+	for _, chain := range zs.externalEvmChain {
+		externalEvmChainConfig, ok := zs.cfg.GetEVMConfig(chain.ChainId)
+		if !ok {
+			return fmt.Errorf("externalEvmChainConfig not found for chain id %d", chain.ChainId)
+		}
+		zetaTokenAddressString := externalEvmChainConfig.ZetaTokenContractAddress
+		zetaTokenAddress := ethcommon.HexToAddress(zetaTokenAddressString)
+		zetatokenNonEth, err := FetchZetaZetaNonEthTokenContract(zetaTokenAddress, zs.evmClient[chain.ChainId])
+		if err != nil {
+			return err
+		}
+		totalSupply, err := zetatokenNonEth.TotalSupply(nil)
+		if err != nil {
+			return err
+		}
+		totalSupplyInt, ok := sdkmath.NewIntFromString(totalSupply.String())
+		if !ok {
+			zs.logger.Error().Msgf("error parsing total supply for chain %d", chain.ChainId)
+			continue
```

Comment:
should we `break` like above when conversion fails?

Reply:
I guess we can continue adding up the supplies to see if the final check passes